Replacing every use of ErrorCall with a custom Exception value

Anton-Latukha opened the pull request:
*(Status: 2020-04-18: finished agenda, awaiting reviews & merge)*

Exceptions for HNix: https://github.com/haskell-nix/hnix/issues/309 "_Replace every use of ErrorCall with a custom Exception value_".

Completed exception values coverage.

For this PR:
1) Keeping current PR as close to `master` implementation as possible.
2) Doing only what was asked in the request, placing groundwork by providing data constructors for exceptions.
3) Decided to keep constructors in groups by the function that exception occurs in. I weighted-in the code structure and what is needed and decided to use this middle way approach.
4) For constructors, the Exception type instances are to process data constructors and it's data into a printable message. With `throwError $ EFunctionNameExceptionName a b c`
5) This is a beauty that exception types are not represented in the signatures, so their data constructors can be rearranged & renamed at any moment into any structure, allowing further evolving the handling.
6) Because the exceptions are grouped by function (3) - they are occurred in, naming convention naturally occurred, you can guess the name of the type that gathers together exceptions you are interested in with, prefix `E*` is for *exception* (*synchronous*), `EA*` for *exception asynchronous*. So it guesses aka `EA` + `FunctionName` (abiding the *camelCalse* rules).

jwiegley commented on src/Nix/Effects/Basic.hs:
```diff
@@ -52,6 +52,14 @@ import           GHC.DataSize
+  deriving Show
+
+
+
```
These new exceptions will need to be rendered, no?

Anton-Latukha replied:
Please, I need a bit more direct information to understand properly.

At this stage structures are drop-in replacements, they are equivalent to strings. In this PR the agenda is to lift the Exception handling from the hardcode into data type constructions, which was requested, allowing use of type class & function handling, which opens ability for the normal Haskell development.

Currently the delta is aimed at to be the first step & to be drop-in straight - using `displayException` produce the `String` that is equivalent to the `String`s earlier hardcoded and do only that. That is why currently data types are so minimal & have only `Show` type class and only it, AFAIK they not to require anything other. We ask the collected info to be `Show, Typeable` and then with it compose `String`, and that is why this constructors need only `Show`, to be an equivalent `String` as currently this was an agenda.

PR is already too large.

After merge I would do a new up to date literature review on the topic, there were valuable articles published, as well as projects created. Then I would open a brief discussion. Then would, as was requested, gradually migrate from `ErrorCall`'s to the found solution.

jwiegley replied:
Thanks, @Anton-Latukha, except that I don't think it's worthwhile to create an intermediate state of development in `master` if the intention to migrate toward real exceptions in the near future. That way, other people working on the project won't get confused as to whether they should use `ErrorCall` or a specific exception.

We have time here, so why don't we just add those commits into this PR branch until it takes on the intended final form.

Anton-Latukha replied:
Ok.

Sorry for the silence for a couple of days.

Ok.

Anton-Latukha committed:
Effects: Basic: add FileError: add CurFileIsntPath

Anton-Latukha committed:
Effects: Basic: FileError: add FileNotInNixPath

Anton-Latukha committed:
Effects: Basic: add NixPathError: add WrongNixPathFormat

Anton-Latukha committed:
Effects: Basic: add FetchTarballError: add NotUrlAttr

Anton-Latukha committed:
Effects: Basic: FetchTarballError: add NorUriNotSet

Anton-Latukha committed:
Effects: Basic: FetchTarballError: add NorUriNorString

Anton-Latukha committed:
Effects: Basic: upd rename Error types to AsyncException types

Ericson2314 commented on src/Nix/Effects/Basic.hs:
```diff
@@ -64,13 +105,7 @@ defaultMakeAbsolutePath origPath = do
-                ++ " __cur_file is in scope,"
-                ++ " but is not a path; it is: "
-                ++ show v
+            v -> throwError $ ErrorCall $ displayException $ CurFileIsntPath v
```
You don't want to call `displayException` here, but where the `m` is eliminated.

Anton-Latukha replied:
Do you mean to make a call out of the `do` block/blocks, or when `-> m FilePath` would be reduced? Surely I am missed something understanding your recommendation.

I would mark your point TODO right now, and dodge it to the further cycle over code. And while then maybe you would write or myself would get more understanding on the Exception process.

Since current approach can be drop-in mass-produced in the code I would first do this.

Anton-Latukha replied:
Your suggestions helped me.

Ericson2314 replied:
Put most simply, there should be no `ErrorCall` in the code base. If you grep for `ErrorCall` you should get nothing.

I'm sorry I didn't put it this way before. I didn't know `ErrorCall` was a type from elsewhere, and not defined in hnix.

Anton-Latukha replied:
Yes.

```haskell
-- | This is thrown when the user calls 'error'. The first @String@ is the
-- argument given to 'error', second @String@ is the location.
data ErrorCall = ErrorCallWithLocation String String
```
currently it seems to be the default constructor for such purposes (currently [`base GHC.Exception`](https://hackage.haskell.org/package/base-4.12.0.0/docs/src/GHC.Exception.html#line-66) considered the main path).

Ericson2314 replied:
Keep in mind throwError comes from the MonadError class and is used purely here (I think). There's no Haskell exceptions from this stuff.

Anton-Latukha replied:
Ok. The Haskell vocabulary and situation is completely perplexing the communication here.
Probably you are saying that the process is pure, until the computation arrives outside into IO monad. Of course - strictly by particularly Haskell theory that it is true, unpurity contained in the IO monad.

`Throw` in combination with `error` terms themselves by the name state that it must end not pure in some cases (Error/Exception or Async Exception/Sync Exception) of terminology I know of.
If throws are not caught - they go into IO monad in the end and trigger the effects. And the majority of situations, of these exception constructors, are not recoverable (unpure), and some are possible to recover with request of IO input (unpure), and the messages are there to be printed/logged, which you guessed...

Yes, the code inside IO monad is effectful and moves into C, for what I know. But if this pure functions are most directly exist to raise the effects - I think if we want we can debate between philosophy, definitions an applications of pure/unpure terms indefinitely in this functions.

Anton-Latukha replied:
But I just basically restated that this is that everyone knows in this thread.

From Haskell terms of pure realm - yes, `throwM` & `MonadThrow` are pure by themselves.
But again - these Exception situations are not recoverable, so their evaluation results into IO effects.

Thanks for raising my attention to this, indeed I was not mentally distinguishing the IO Rubicon line.

Ericson2314 replied:
https://hackage.haskell.org/package/mtl-2.2.2/docs/Control-Monad-Except.html#v:throwError see the instances for that class. there is the IOException instance which uses exceptions but I don't think we are using that.

Anton-Latukha replied:
The `throwError` is redefined in HNix:
https://github.com/haskell-nix/hnix/blob/c03a40aceac5221c82603c9208b381a76c72100f/src/Nix/Frames.hs#L53-L58

GHCI:
```haskell
*Nix> :l src/Nix.hs
...
*Nix> :i throwError
throwError ::
  (Framed e m, Exception s, Control.Monad.Catch.MonadThrow m) =>
  s -> m a
  	-- Defined at src/Nix/Frames.hs:55:1
```
`toException` [is](https://hackage.haskell.org/package/base-4.12.0.0/docs/src/GHC.Exception.Type.html#toException):
```haskell
class (Typeable e, Show e) => Exception e where
    toException   :: e -> SomeException
    fromException :: SomeException -> Maybe e

    toException = SomeException
    fromException (SomeException e) = cast e
```

& when `throwM` is launched - it shortens the monadic computation.

There is a bunch of custom HNix type classes & instances extending the IO monad in `src/Nix/Effects.hs`:
https://github.com/haskell-nix/hnix/blob/c03a40aceac5221c82603c9208b381a76c72100f/src/Nix/Effects.hs#L86-L113

`Effects.hs` has a second largest Exception counter over all HNix files, I am moving towards it. Also it is the root of all evil, so I decided to learn the simplier, more isolated, abstracted modules before.

I just already learned a bunch of project code while working on it. Also normal tooling fitting to the project helps here: https://github.com/Anton-Latukha/haskell-notes#integration-of-nixosnix-with-haskell-ide-engine-hie-and-emacs-spacemacs.

Anton-Latukha replied:
Also I see (as also you can see here) there is too much forward and backward transformations of Exceptions, these computations seems to be can be reduced.

Ericson2314 replied:
My apologies I didn't realize we were doing those things. @jwiegley has the final say, but I would ditch Exception and SomeException and only use closed error types. I really don't like open error types.

Anton-Latukha replied:
No apologies, as I do errors everyone does them.

Currently I'm just doing a bunch of Exception constructors as was asked.
I am gathering a list of major point questions to ask/tweak during the merge preparation.

Ericson2314 replied:
In that case you will want to implement the `Exception` class so you can `throwError $ YourNewNiceType arg0 ...argn` directly without any `ErrorCall`.

jwiegley replied:
@Ericson2314 is right, instead of `ErrorCall $ displayException $ EMonadExecMissingProgram u u`, it should just be `EMonadExecMissingProgram u u`. The rendering into a string part should only happen at reporting time.

Anton-Latukha replied:
Yep, reduced that.

Anton-Latukha committed:
Effects: Basic: defaultMakeAbsolutePath: add TODO

Anton-Latukha committed:
Eval: add AttrSetAlterInvalidSelector exception

Anton-Latukha committed:
Eval: add ComponentValueIsNullExpectedString exception

Anton-Latukha committed:
Eval: add BuildArgumetMissingValue exception

Anton-Latukha committed:
Eval: add BuildArgumentUnexpectedParameter exception

Anton-Latukha committed:
Coerse: add CoerseUnexpectedArgument exception

Anton-Latukha committed:
Effects: Basic: add TODO to not processed exception throws

Anton-Latukha committed:
Render: Frame: add FrameAsyncException: add UnrecognizedFrame

rycee commented on src/Nix/String/Coerce.hs:
```diff
@@ -38,6 +38,15 @@ data CopyToStoreMode
   -- ^ Add paths to the store as they are encountered
   deriving (Eq,Ord,Enum,Bounded)
 
+newtype CoerseAsyncException a
```
`CoerseAsyncException` → `CoerceAsyncException`

Anton-Latukha replied:
Thank you, `rycee`, you already helped me several times in the community.

Anton-Latukha committed:
Lint: add MergeAsyncException: add MergeClosuresException

Anton-Latukha committed:
Lint: MergeAsyncException: add MergeBuiltinsException

Anton-Latukha committed:
Lint: MergeAsyncException: add MergeImpossibleException

Anton-Latukha committed:
Lint: add UnifyAsyncException: add UnifyNotDoneException

Anton-Latukha committed:
Lint: UnifyAsyncException: add UnifyUnexpectedCaseException

Anton-Latukha committed:
Lint: upd minor fxs

Anton-Latukha committed:
Lint: add MonadEvalAsyncException: add MonadEvalAttrUnknownInheritException

Anton-Latukha committed:
Lint: MonadEvalAsyncException: add MonadEvalAttrNotFound TODO notes

Anton-Latukha committed:
Lint: MonadEvalAsyncException: add MonadEvalNotImplementedException

Anton-Latukha committed:
Lint: MonadEvalAsyncException: add MonadEvalScopeNotASetWithStatementException

Anton-Latukha committed:
Lint: add LintAppAsyncException: add LintAppNotFuncException

Anton-Latukha committed:
Lint: LintAppAsyncException: add LintAppNotImplementedException

Anton-Latukha committed:
Lint: LintAppAsyncException: add LintAppNotImplementedNManyNotSetException

Anton-Latukha committed:
Lint: LintAppAsyncException: add LintAppNotImplementedBuiltinException

Anton-Latukha committed:
Lint: LintAppAsyncException: add LintAppNotImplementedSetException

Anton-Latukha committed:
Lint: LintAppAsyncException: add LintAppCallNotFunctionException

Anton-Latukha committed:
Lint: add MonadCatchAsyncException: add MonadCatchCanNotCatchInLintException

Anton-Latukha committed:
upd abbreviate *(Exception -> E) in type constructors; add docs on `E`

Anton-Latukha committed:
Effects: add MonadExecAsyncE: add MonadExecMissingProgramE

Anton-Latukha committed:
Effects: MonadExecAsyncE: add MonadExecNoOutputE

Anton-Latukha committed:
Effects: MonadExecAsyncE: add MonadExecParsingOutputE

Anton-Latukha committed:
Effects: MonadExecAsyncE: add MonadExecFailE

Anton-Latukha committed:
Effects: add MonadInstantiateAsuncE: add MonadInstantiateParsingOutputE

Anton-Latukha committed:
Effects: MonadInstantiateAsyncE: add MonadInstantiateFailE

Anton-Latukha committed:
Effects: add MonadHttpAsyncE: add MonadHttpFetchFailE

Anton-Latukha committed:
Effects: MonadHttpAsyncE: add MonadHttpStoreNotReady

Anton-Latukha committed:
Effects: add MonadStoreAsyncE: add MonadStoreAddPathFailE

Anton-Latukha committed:
Lint: upd fx data constructors *Exception -> *E

Anton-Latukha committed:
upd exception constructors names: data *E -> E*, type *AsyncE -> EA*

Anton-Latukha committed:
README-design: upd docs on exception constructors naming

Anton-Latukha committed:
Exec: add EAMonadNix: add EMonadNixUndefinedVar

Anton-Latukha committed:
Exec: EAMonadNix: add EMonadNixUnknownAttrInherit

Anton-Latukha committed:
Exec: EAMonadNix: add EMonadNixAttrNotFound

Anton-Latukha committed:
Exec: EAMonad: add EMonadNixStringAssembleFail

Anton-Latukha committed:
Exec: add EACallFunc: add ECallFuncCallStackExausted

Anton-Latukha committed:
Exec: EACallFunc: add ECallFuncCalledNotFunction

Anton-Latukha committed:
Exec: add EAExecUnaryOp: add EExecUnaryOpUnsupportedType

Anton-Latukha committed:
Exec: EAExecUnaryOp: add EExecUnaryOpEvaluatedToNotAtomicType

Anton-Latukha committed:
Exec: add EAFromStringNoContext: add EFromStringNoContext

Anton-Latukha committed:
Exec: EAMonadEval: rname EAMonad{Nix -> Eval), add polymorphism, upd

Anton-Latukha committed:
Exec: EAExecUnaryOp: add polymorphism

Anton-Latukha committed:
Exec: add EAUnsupportedTypes

Anton-Latukha committed:
Exec: execBinaryOpForced: upd unsupportedTypes

Here is a tricky type inference that I could not gather into one.
For `compare` created separate hacky function that needs to be merged-in

Anton-Latukha committed:
Exec: add EAAlreadyHandled; execBinaryOpForced: upd alreadyHandled

Anton-Latukha committed:
Exec: add EAExecBinaryOpForced: add EExecBinaryOpForcedNPlusStringToPath

Anton-Latukha committed:
Exec: EAExecBinaryOpForced: upd comment link to permalink

State-dependent link. Comment is two month old, and already not useful.

Doing things proper - here is permalink.

Anton-Latukha committed:
Exec: EAExecBinaryOpForced: merge-in the EAUnsuportedTypes

Anton-Latukha committed:
Exec: EAExecBinaryOpForced: merge-in the EAAlreadyHandled

Anton-Latukha committed:
Exec: EAExecBinaryOpForced: add EExrcBinaryOpForcedNApp

Anton-Latukha committed:
Basic: add EADefaultImportPath: add EDefaultImportPathParse

Anton-Latukha committed:
Basic: EADefaultImportPath: add EDefaultImportPath

Anton-Latukha committed:
Builtins: add EAMkThunk

Anton-Latukha committed:
Builtins: add EAFoldNixPath

Anton-Latukha committed:
Builtins: add EAAttrsetGet

Anton-Latukha committed:
Builtins: add EAUnsafeGetAttrPos

Anton-Latukha committed:
Builtins: add EAHead_

Anton-Latukha committed:
Builtins: add EATail_

Anton-Latukha committed:
Builtins: reformat fun substring

Anton-Latukha committed:
Builtins: add EASubString

Anton-Latukha committed:
Builtins: add EAMap_

Anton-Latukha committed:
Builtins: add EAMapAttrs_

Anton-Latukha committed:
Builtins: add EADirOf

Anton-Latukha committed:
Builtins: add EAElemAt_

Anton-Latukha committed:
Builtins: add EAGenList

Anton-Latukha committed:
Builtins: add EAGenericClosure

Anton-Latukha committed:
Builtins: add EAReplaceStrings

Anton-Latukha committed:
Builtins: add EAPathExists_

Anton-Latukha committed:
Builtins: add EAFunctionArgs

Anton-Latukha committed:
Builtins: add EALessThan

Anton-Latukha committed:
Builtins: add EAHashString

Anton-Latukha committed:
Builtins: add EAAbsolutePathFromValue

Anton-Latukha committed:
Builtins: add EAFindFile_

Anton-Latukha committed:
Builtins: add EAFromJSON

Anton-Latukha committed:
Builtins: add EAFetchurl

Anton-Latukha committed:
Builtins: add EAGetContext

Anton-Latukha committed:
Builtins: add EAAppendContext

jwiegley reviewed (changes requested):
See feedback in the currently open comment.

Anton-Latukha committed:
Builtins: upd builtinsList: reduce ErrorCall in mkThunk

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: foldNixPath: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: attrsetGet: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: unsafeGetAttrPos: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: head_: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: tail_: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: prettify substring

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: map_: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: mapAttrs_: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: dirOf: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: elemAt_: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: genList: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: genericClosure: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: genericClosure: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: genericClosure: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: replaceStrings: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: functionArgs: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: pathExists_: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: lessThanUnsupportedArgs: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: hashStringUnsupportedValue: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: absolutePathFromValue: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: absolutePathFromValue: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: findFile_: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: findFile_: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: findFile_: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: fromJSON: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: fetchurl: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: fetchurl: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: fetchurl: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: getContext: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: appendContext: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: appendContext: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Builtins: upd: appendContext: reduce ErrorCall

M  src/Nix/Builtins.hs

Anton-Latukha committed:
Effects: class MonadInstantiate: instantiateExpr: prettify

M  src/Nix/Effects.hs

Anton-Latukha committed:
Effects/Basic: upd: defaultMakeAbsolutePath: reduce ErrorCall

M  src/Nix/Effects/Basic.hs

Anton-Latukha committed:
Effeects/Basic: upd: findPathBy: reduce ErrorCall

M  src/Nix/Effects/Basic.hs

Anton-Latukha committed:
Effeects/Basic: split EAFile into types by function of exception origin

M  src/Nix/Effects/Basic.hs

Anton-Latukha committed:
Effeects/Basic: upd: findPathBy: mrg-in except constr, reduce ErrorCall

M  src/Nix/Effects/Basic.hs

Anton-Latukha committed:
Effeects/Basic: upd: fetchTarball: reduce ErrorCall

M  src/Nix/Effects/Basic.hs

Anton-Latukha committed:
Effeects/Basic: fetchTarball: rname data constrs, reduce ErrorCalls

M  src/Nix/Effects/Basic.hs

Anton-Latukha committed:
Effeects/Basic: upd: defaultImportPath: reduce ErrorCall

M  src/Nix/Effects/Basic.hs

Anton-Latukha committed:
Effeects/Basic: upd: defaultImportPath: reduce ErrorCall

M  src/Nix/Effects/Basic.hs

Anton-Latukha committed:
Eval: attrSetAlter: reduce ErrorCall

M  src/Nix/Eval.hs

Anton-Latukha committed:
Eval: attrSetAlter: prettify

M  src/Nix/Eval.hs

Anton-Latukha committed:
Eval: EAAttrSetAlter: prettify

M  src/Nix/Eval.hs

Anton-Latukha committed:
Eval: evalGetterKeyName: reduce ErrorCall

M  src/Nix/Eval.hs

Anton-Latukha committed:
Eval: buildArgument: reduce ErrorCall

M  src/Nix/Eval.hs

Anton-Latukha committed:
Eval: buildArgument: reduce ErrorCall

M  src/Nix/Eval.hs

Anton-Latukha committed:
Exec: inst MonadEval (NValue t f m) m: freeVariable: reduce ErrorCall

M  src/Nix/Exec.hs

Anton-Latukha committed:
Exec: inst MonadEval (NValue t f m) m: attrMissing: reduce ErrorCall

M  src/Nix/Exec.hs

Anton-Latukha committed:
Exec: inst MonadEval (NValue t f m) m: attrMissing: reduce ErrorCall

M  src/Nix/Exec.hs

Anton-Latukha committed:
Exec: inst MonadEval (NValue t f m) m: evalString: reduce ErrorCall

M  src/Nix/Exec.hs

Anton-Latukha committed:
Exec: callFunc: reduce ErrorCall

M  src/Nix/Exec.hs

Anton-Latukha committed:
Exec: callFunc: reduce ErrorCall

M  src/Nix/Exec.hs

Anton-Latukha committed:
Exec: execUnaryOp: reduce ErrorCall

M  src/Nix/Exec.hs

Anton-Latukha committed:
Exec: execUnaryOp: reduce ErrorCall

M  src/Nix/Exec.hs

Anton-Latukha committed:
Exec: execBinaryOp: reduce ErrorCall

M  src/Nix/Exec.hs

Anton-Latukha committed:
Exec: execBinaryOp: reduce ErrorCall

M  src/Nix/Exec.hs

Anton-Latukha committed:
Exec: execBinaryOpForced: prettify

M  src/Nix/Exec.hs

Anton-Latukha committed:
Exec: execBinaryOpForced: reduce ErrorCall

M  src/Nix/Exec.hs

Anton-Latukha committed:
Exec: execBinaryOpForced: reduce ErrorCall

M  src/Nix/Exec.hs

Anton-Latukha committed:
Exec: execBinaryOpForced: reduce ErrorCall

M  src/Nix/Exec.hs

Anton-Latukha committed:
Exec: fromStringNoContext: reduce ErrorCall

M  src/Nix/Exec.hs

Anton-Latukha committed:
Json: nvalueToJSON: prettify

M  src/Nix/Json.hs

Anton-Latukha committed:
Lint: merge: reduce ErrorCall

M  src/Nix/Lint.hs

Anton-Latukha committed:
Lint: merge: reduce ErrorCall

M  src/Nix/Lint.hs

Anton-Latukha committed:
Lint: unify: reduce ErrorCall

M  src/Nix/Lint.hs

Anton-Latukha committed:
Lint: inst MonadEval: attrMissing: reduce ErrorCall

M  src/Nix/Lint.hs

Anton-Latukha committed:
Lint: inst MonadEval: evalWith: reduce ErrorCall

M  src/Nix/Lint.hs

Anton-Latukha committed:
Lint: lintApp: reduce ErrorCall

M  src/Nix/Lint.hs

Anton-Latukha committed:
Lint: lintApp: reduce ErrorCall

M  src/Nix/Lint.hs

Anton-Latukha committed:
Lint: lintApp: reduce ErrorCall

M  src/Nix/Lint.hs

Anton-Latukha committed:
Lint: lintApp: reduce ErrorCall

M  src/Nix/Lint.hs

Anton-Latukha committed:
Lint: lintApp: reduce ErrorCall

M  src/Nix/Lint.hs

Anton-Latukha committed:
String: principleMakeNixString: prettify

M  src/Nix/String.hs

Anton-Latukha committed:
String/Coerce: coerceToString: reduce ErrorCall

M  src/Nix/String/Coerce.hs

Anton-Latukha committed:
String/Coerce: rname EACoerce(->ToString) and its data constructor

M  src/Nix/String/Coerce.hs

Anton-Latukha committed:
README-design: Exception types: upd defaulting to syncronous

M  README-design.md

Anton-Latukha committed:
TOTAL: rename all Exception constructors as (they are) - synchronous

M  src/Nix/Builtins.hs
M  src/Nix/Effects.hs
M  src/Nix/Effects/Basic.hs
M  src/Nix/Eval.hs
M  src/Nix/Exec.hs
M  src/Nix/Lint.hs
M  src/Nix/Render/Frame.hs
M  src/Nix/String/Coerce.hs
